Jenkins distribution customize service proposal (#2813)
jenkins-infra/jenkins.io: content/projects/gsoc/2020/project-ideas/jenkins-distribution-customize-service.adoc

---
layout: gsocprojectidea

> title: "Jenkins distribution customize service"

Reviewer: ```suggestion
title: "Custom Jenkins distribution build service"
```

> goal: "Provide an out of the box solution for Jenkins users"

Reviewer: ```suggestion
goal: "Provide an out of the box solution for packaging Jenkins distributions as WAR files or Docker images"
```

Reviewer: If you consider other output types, I suggest adding them explicitly

Author: Good point. Consider the scope of GSoC, I think that two types are enough.

> category: Service

Reviewer: ```suggestion
category: Tools
```

No strong opinion

year: 2020
status: draft
skills:
- Java

> - Web Development

Reviewer: ```suggestion
- Web Development
- Docker
```

mentors:
- "linuxsuren"
---

> For many users, they download Jenkins first, then select some plugins and config them. 

Reviewer: ```suggestion
For many users, they download Jenkins first, then select some plugins and configure them. 
```

> It might takes a lot of time, like hours. But if we can get a perfect Jenkins distribution which contains all we need, 

Reviewer 2: ```suggestion
It might take a lot of time, like hours. But if we can get a perfect Jenkins distribution which contains all we need, 
```

it can save that time for us.

> The service could like this, a website offered as https://customize.jenkins.io. Plus, it should support self-host. 

Reviewer 2: ```suggestion
The service could be in the form of  a website offered as https://customize.jenkins.io. Plus, it should support self-host. 
```

People can select the following configuration items:

* Jenkins core version
* plugins

> * common configuration, user/password, update-center site e.g.

Reviewer: How? JCasC? Or would it be dialog-based?

Author: I'd prefer it's scenario-based instead of using JCasC directly. I hope it can be very easy to use. Even for some new users, they don't need to know too much about Jenkins to start it.

Reviewer: If we do that, it would be great to share the implementation with the Jenkins Installation Wizard

Author: +1 from me

* plugin based configuration, Kubernetes, Sonarqube plugin config e.g.
* multi-form package, jenkins.war or docker image

> * other things

Reviewer 3: How about a backup/restore solution?

Reviewer 3: Also, where to store Jenkins logs might interesting

Author: Agree. Other options also can be considered. https://jenkins.io/sigs/cloud-native/pluggable-storage/

Reviewer 2: https://github.com/jenkinsci/aws-cloudwatch-logs-publisher-plugin

> Two projects are expected backend and frontend. Thanks to link:https://github.com/jenkinsci/custom-war-packager[Custom Jenkins WAR packager for Jenkins], 

Reviewer: ```suggestion
The project is expected to have a backend and a frontend parts.
Thanks to link:https://github.com/jenkinsci/custom-war-packager[Custom WAR Packager for Jenkins], 
```

> it already did a lot of works. It can be the main library for the backend project. 

Reviewer 2: ```suggestion
it already does a lot of the work . It can be the main library for the backend project. 
```

Author: That project has done a lot of the work. So I use `did` instead of `does`. Please correct me if I'm wrong.

Reviewer 3: Not sure how nit picky we want to be in English grammar, but in that case, "it already did a lot of the work" would be correct

Reviewer 2: Yeah not a problem :smile: just a suggestion

> The backend should provide the modern Restful API. We can develop different client base on that, UI or CLI. A readable API document is

Reviewer 2: ```suggestion
The backend should provide the modern Restful API. We can develop different clients based on that ie: UI or CLI. A readable API document is
```

> The backend should provide the modern Restful API. We can develop different clients based on that ie: UI or CLI. A readable API document is

Reviewer: ```suggestion
The backend should provide a modern Restful API.
We can develop different clients based on it, e.g. a Command Line Interface or a Desktop GUI application.. 
A readable API documentation is
```

necessary. Like many projects, the Swagger UI is a good example.

> We can start the backend project with SpringBoot, it might save a lot of time.

Reviewer 2: One of the things to consider is a front end UI, could be react or even angular.

Author: I don't have enough experience with the front-end. Any ideas are welcome.

Reviewer 3: Blue Ocean was using React, so as of last year, that would have seemed like the right choice. I'm not sure what's up with Blue Ocean these days, so maybe there'll be a new UI framework?

Ignoring that choice, React seems to be more popular, and also newer FWIW. We switched all of our front end from Angular to React a couple years back.

Reviewer 2: Yeah React/Angular seem to be the most popular choices for a frontend. Some pros in favour of react (Personal Experience here :) ) 

1. Single Way DataFlow ( Events Percolate upwards. Something which we need) .
2. Testablility
3. Huge Community Support.
4. Performance

## References

Below you can find a list of existing plugins and libraries which can be used for inspiration:

> * link:https://github.com/jenkinsci/custom-war-packager[Custom Jenkins WAR packager for Jenkins]

Author: I copy it from here 
![image](https://user-images.githubusercontent.com/1450685/73583172-abd2f300-44cb-11ea-947c-e132f297f975.png)
I guess you might want to change it too.

---
layout: gsocprojectidea
title: "Custom Jenkins distribution build service"
goal: "Provide an out of the box solution for packaging Jenkins distributions as WAR files or Docker images"
category: Tools
year: 2020
status: draft
skills:
- Java
- Web Development
- Docker
mentors:
- "linuxsuren"
links:

>   gitter: "jenkinsci/custom-distribution-build-service"

Reviewer: rename it to match the proposal? We could also use an existing entity like Platform SIG for now

Author: Agree, good idea. Let's reuse the Platform SIG.
